File: playwright/update-bank.js
/**
 * FlowAccount – อัปเดตข้อมูลธนาคารของ Contacts จาก Google Sheet
 *
 * วิธีใช้:
 *   1) cd playwright && npm install
 *   2) คัดลอก .env.example -> .env แล้วใส่ email/password ของ FlowAccount
 *   3) ตรวจสอบว่า Google Sheet ตั้งค่า share = "ใครก็ตามที่มีลิงก์ดูได้"
 *   4) npx playwright install chromium  (ครั้งแรกเท่านั้น)
 *   5) node update-bank.js
 *
 * หมายเหตุ:
 *   - สคริปต์เปิด browser แบบมีหน้าจอ (headless: false) เพื่อให้ผู้ใช้
 *     ทำ 2FA / CAPTCHA ได้เอง ถ้าเจอ ให้ทำเอง แล้วสคริปต์จะรันต่อ
 *   - Selector ของ FlowAccount เป็นการเดาแบบ best-effort
 *     ถ้า UI เปลี่ยน ให้แก้ตรง section "SELECTORS" ด้านล่าง
 */

const { chromium } = require('playwright');
const https = require('https');
require('dotenv').config();

// ---------- CONFIG ----------
const SHEET_ID = '1CR5T5XnyFSIT1r2q-XKuxowOk4IiQTl_dLVM0aJe-vs';
const CSV_URL = `https://docs.google.com/spreadsheets/d/${SHEET_ID}/export?format=csv`;
const CONTACTS_URL = 'https://advance.flowaccount.com/N4236690/business/contacts';
const LOGIN_URL = 'https://flowaccount.com/login';

const EMAIL = process.env.FLOWACCOUNT_EMAIL;
const PASSWORD = process.env.FLOWACCOUNT_PASSWORD;

const BRANCH_CODE = '0000';
const BRANCH_NAME = '-';
const ACCOUNT_TYPE = 'ออมทรัพย์';

// ---------- SELECTORS (แก้ตรงนี้ถ้า UI FlowAccount เปลี่ยน) ----------
const SEL = {
  loginEmail: 'input[type="email"], input[name="email"], input[name="username"]',
  loginPassword: 'input[type="password"], input[name="password"]',
  loginSubmit: 'button[type="submit"], button:has-text("เข้าสู่ระบบ"), button:has-text("Login")',

  contactsSearch: 'input[type="search"], input[placeholder*="ค้นหา"], input[placeholder*="Search"]',
  contactRowByName: (name) => `tr:has-text("${name}"), [role="row"]:has-text("${name}"), a:has-text("${name}")`,

  editButton: 'button:has-text("แก้ไข"), a:has-text("แก้ไข"), button[aria-label*="แก้ไข"]',
  bankTab: 'button:has-text("ข้อมูลธนาคาร"), a:has-text("ข้อมูลธนาคาร"), [role="tab"]:has-text("ธนาคาร"), button:has-text("บัญชีธนาคาร")',
  addBankButton: 'button:has-text("เพิ่มบัญชีธนาคาร"), button:has-text("เพิ่ม")',

  bankNameField:    'input[name*="bankName"], input[placeholder*="ธนาคาร"], [aria-label*="ธนาคาร"]',
  accountNameField: 'input[name*="accountName"], input[placeholder*="ชื่อบัญชี"]',
  accountNumberField: 'input[name*="accountNumber"], input[placeholder*="เลขที่บัญชี"]',
  branchCodeField:  'input[name*="branchCode"], input[placeholder*="รหัสสาขา"]',
  branchNameField:  'input[name*="branchName"], input[placeholder*="ชื่อสาขา"]',
  accountTypeField: 'select[name*="accountType"], [aria-label*="ประเภทบัญชี"], input[placeholder*="ประเภทบัญชี"]',

  saveAndClose: 'button:has-text("บันทึกแล้วปิด")',
};

// ---------- HELPERS ----------
function fetchCsv(url, redirects = 0) {
  return new Promise((resolve, reject) => {
    if (redirects > 5) return reject(new Error('Too many redirects'));
    https.get(url, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return resolve(fetchCsv(res.headers.location, redirects + 1));
      }
      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode} – ตรวจสอบว่า Sheet share = Anyone with the link`));
      }
      let data = '';
      res.setEncoding('utf8');
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cur = '';
  let inQuote = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (inQuote) {
      if (ch === '"' && text[i + 1] === '"') { cur += '"'; i++; }
      else if (ch === '"') inQuote = false;
      else cur += ch;
    } else {
      if (ch === '"') inQuote = true;
      else if (ch === ',') { row.push(cur); cur = ''; }
      else if (ch === '\n') { row.push(cur); rows.push(row); row = []; cur = ''; }
      else if (ch === '\r') { /* skip */ }
      else cur += ch;
    }
  }
  if (cur.length || row.length) { row.push(cur); rows.push(row); }
  return rows;
}

async function safeFill(page, selector, value) {
  const el = page.locator(selector).first();
  await el.waitFor({ state: 'visible', timeout: 8000 });
  await el.fill('');
  await el.fill(value);
}

async function safePickBank(page, bankName) {
  // ฟิลด์ธนาคารใน FlowAccount มักเป็น combobox/dropdown
  const field = page.locator(SEL.bankNameField).first();
  await field.waitFor({ state: 'visible', timeout: 8000 });
  await field.click();
  await field.fill(bankName);
  await page.waitForTimeout(800);
  // เลือกตัวเลือกที่ตรงในเมนู
  const option = page.locator(`[role="option"]:has-text("${bankName}"), li:has-text("${bankName}")`).first();
  if (await option.count()) await option.click();
}

async function safePickAccountType(page, label) {
  const field = page.locator(SEL.accountTypeField).first();
  await field.waitFor({ state: 'visible', timeout: 8000 });
  const tag = await field.evaluate((el) => el.tagName.toLowerCase());
  if (tag === 'select') {
    await field.selectOption({ label });
  } else {
    await field.click();
    await page.waitForTimeout(400);
    const opt = page.locator(`[role="option"]:has-text("${label}"), li:has-text("${label}")`).first();
    if (await opt.count()) await opt.click();
    else await field.fill(label);
  }
}

// ---------- MAIN ----------
(async () => {
  if (!EMAIL || !PASSWORD) {
    console.error('ERROR: ตั้งค่า FLOWACCOUNT_EMAIL และ FLOWACCOUNT_PASSWORD ใน .env');
    process.exit(1);
  }

  console.log('โหลด Google Sheet...');
  const csv = await fetchCsv(CSV_URL);
  const rows = parseCsv(csv);
  if (rows.length < 2) {
    console.error('Sheet ว่างหรืออ่านไม่ได้');
    process.exit(1);
  }

  // skip header row
  const records = rows.slice(1).map((r) => ({
    accountName:   (r[0]  || '').trim(),  // Column A
    bankName:      (r[17] || '').trim(),  // Column R
    accountNumber: (r[18] || '').trim(),  // Column S
  })).filter((x) => x.accountName && x.bankName && x.accountNumber);

  console.log(`เจอ ${records.length} รายการที่มีข้อมูลธนาคารครบ`);

  const browser = await chromium.launch({ headless: false, slowMo: 80 });
  const context = await browser.newContext({ locale: 'th-TH' });
  const page = await context.newPage();

  // -- LOGIN --
  console.log('กำลัง login FlowAccount...');
  await page.goto(LOGIN_URL, { waitUntil: 'domcontentloaded' });
  try {
    await safeFill(page, SEL.loginEmail, EMAIL);
    await safeFill(page, SEL.loginPassword, PASSWORD);
    await page.locator(SEL.loginSubmit).first().click();
  } catch (e) {
    console.warn('Auto-login อาจไม่สำเร็จ — กรุณา login เอง แล้วสคริปต์จะรันต่อ');
  }

  // รอจน URL เข้าหน้าหลัก (เผื่อ 2FA)
  console.log('รอเข้าสู่ระบบเสร็จ (รวม 2FA ถ้ามี)... สูงสุด 3 นาที');
  await page.waitForURL(/flowaccount\.com\/N\d+/, { timeout: 180000 }).catch(() => {});

  const results = { ok: [], failed: [] };

  for (const rec of records) {
    console.log(`\n→ ${rec.accountName} | ${rec.bankName} | ${rec.accountNumber}`);
    try {
      await page.goto(CONTACTS_URL, { waitUntil: 'domcontentloaded' });
      await page.waitForLoadState('networkidle').catch(() => {});

      // ค้นหา contact
      const search = page.locator(SEL.contactsSearch).first();
      await search.waitFor({ state: 'visible', timeout: 15000 });
      await search.fill('');
      await search.fill(rec.accountName);
      await page.waitForTimeout(1500);

      // คลิกผลลัพธ์แรกที่ตรง
      const target = page.locator(SEL.contactRowByName(rec.accountName)).first();
      await target.waitFor({ state: 'visible', timeout: 10000 });
      await target.click();
      await page.waitForLoadState('networkidle').catch(() => {});

      // กดปุ่ม "แก้ไข"
      const editBtn = page.locator(SEL.editButton).first();
      if (await editBtn.count()) {
        await editBtn.click();
        await page.waitForLoadState('networkidle').catch(() => {});
      }

      // ไป tab/section "ข้อมูลธนาคาร"
      const bankTab = page.locator(SEL.bankTab).first();
      if (await bankTab.count()) {
        await bankTab.click();
        await page.waitForTimeout(500);
      }

      // ถ้ามีปุ่ม "เพิ่มบัญชีธนาคาร" ให้กด
      const addBtn = page.locator(SEL.addBankButton).first();
      if (await addBtn.count()) {
        await addBtn.click().catch(() => {});
        await page.waitForTimeout(500);
      }

      // กรอกข้อมูล
      await safePickBank(page, rec.bankName);
      await safeFill(page, SEL.accountNameField, rec.accountName);
      await safeFill(page, SEL.accountNumberField, rec.accountNumber);
      await safeFill(page, SEL.branchCodeField, BRANCH_CODE);
      await safeFill(page, SEL.branchNameField, BRANCH_NAME);
      await safePickAccountType(page, ACCOUNT_TYPE);

      // กดบันทึกแล้วปิด
      await page.locator(SEL.saveAndClose).first().click();
      await page.waitForLoadState('networkidle').catch(() => {});
      console.log('  ✓ บันทึกแล้ว');
      results.ok.push(rec.accountName);
    } catch (e) {
      console.error(`  ✗ ล้มเหลว: ${e.message}`);
      results.failed.push({ name: rec.accountName, error: e.message });
      // ถ่ายภาพหน้าจอเก็บไว้ debug
      await page.screenshot({ path: `error_${Date.now()}.png`, fullPage: true }).catch(() => {});
    }
  }

  console.log('\n========== สรุป ==========');
  console.log(`สำเร็จ: ${results.ok.length}`);
  console.log(`ล้มเหลว: ${results.failed.length}`);
  if (results.failed.length) console.log(JSON.stringify(results.failed, null, 2));

  await browser.close();
})();

File: selenium/update_bank.py
"""
FlowAccount - อัปเดตข้อมูลธนาคารของ Contacts จาก Google Sheet (Selenium)

วิธีใช้:
  1) cd selenium
  2) python -m venv .venv && source .venv/bin/activate   # หรือ .venv\Scripts\activate บน Windows
  3) pip install -r requirements.txt
  4) คัดลอก .env.example -> .env แล้วใส่ email/password ของ FlowAccount
  5) ตรวจสอบว่า Google Sheet share = "ใครก็ตามที่มีลิงก์ดูได้"
  6) python update_bank.py

หมายเหตุ:
  - เปิด Chrome แบบมีหน้าจอ (ไม่ headless) เผื่อมี 2FA / CAPTCHA
  - Selector เป็นการเดาแบบ best-effort -- ถ้า UI FlowAccount เปลี่ยน
    แก้ตรง class SELECTORS ด้านล่างได้เลย
"""

import csv
import io
import os
import sys
import time
import urllib.request

from dotenv import load_dotenv
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import Select, WebDriverWait

load_dotenv()

# ---------- CONFIG ----------
SHEET_ID = "1CR5T5XnyFSIT1r2q-XKuxowOk4IiQTl_dLVM0aJe-vs"
CSV_URL = f"https://docs.google.com/spreadsheets/d/{SHEET_ID}/export?format=csv"
CONTACTS_URL = "https://advance.flowaccount.com/N4236690/business/contacts"
LOGIN_URL = "https://flowaccount.com/login"

EMAIL = os.getenv("FLOWACCOUNT_EMAIL")
PASSWORD = os.getenv("FLOWACCOUNT_PASSWORD")

BRANCH_CODE = "0000"
BRANCH_NAME = "-"
ACCOUNT_TYPE = "ออมทรัพย์"


# ---------- SELECTORS (แก้ตรงนี้ถ้า UI FlowAccount เปลี่ยน) ----------
class SELECTORS:
    LOGIN_EMAIL = (By.CSS_SELECTOR, 'input[type="email"], input[name="email"], input[name="username"]')
    LOGIN_PASSWORD = (By.CSS_SELECTOR, 'input[type="password"], input[name="password"]')
    LOGIN_SUBMIT = (By.XPATH, '//button[@type="submit" or contains(., "เข้าสู่ระบบ") or contains(., "Login")]')

    CONTACT_SEARCH = (By.CSS_SELECTOR, 'input[type="search"], input[placeholder*="ค้นหา"], input[placeholder*="Search"]')

    EDIT_BUTTON = (By.XPATH, '//button[contains(., "แก้ไข")] | //a[contains(., "แก้ไข")]')
    BANK_TAB = (By.XPATH, '//*[self::button or self::a or @role="tab"][contains(., "ข้อมูลธนาคาร") or contains(., "บัญชีธนาคาร")]')
    ADD_BANK_BUTTON = (By.XPATH, '//button[contains(., "เพิ่มบัญชีธนาคาร") or normalize-space()="เพิ่ม"]')

    BANK_NAME_FIELD = (By.CSS_SELECTOR, 'input[name*="bankName"], input[placeholder*="ธนาคาร"], [aria-label*="ธนาคาร"]')
    ACCOUNT_NAME_FIELD = (By.CSS_SELECTOR, 'input[name*="accountName"], input[placeholder*="ชื่อบัญชี"]')
    ACCOUNT_NUMBER_FIELD = (By.CSS_SELECTOR, 'input[name*="accountNumber"], input[placeholder*="เลขที่บัญชี"]')
    BRANCH_CODE_FIELD = (By.CSS_SELECTOR, 'input[name*="branchCode"], input[placeholder*="รหัสสาขา"]')
    BRANCH_NAME_FIELD = (By.CSS_SELECTOR, 'input[name*="branchName"], input[placeholder*="ชื่อสาขา"]')
    ACCOUNT_TYPE_FIELD = (By.CSS_SELECTOR, 'select[name*="accountType"], [aria-label*="ประเภทบัญชี"], input[placeholder*="ประเภทบัญชี"]')

    SAVE_AND_CLOSE = (By.XPATH, '//button[contains(., "บันทึกแล้วปิด")]')


# ---------- HELPERS ----------
def fetch_csv(url: str) -> list[list[str]]:
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req) as resp:
        if resp.status != 200:
            raise RuntimeError(f"HTTP {resp.status} - ตรวจสอบว่า Sheet share = Anyone with the link")
        data = resp.read().decode("utf-8", errors="replace")
    return list(csv.reader(io.StringIO(data)))


def safe_fill(driver, by_selector, value: str, timeout: int = 8):
    el = WebDriverWait(driver, timeout).until(EC.visibility_of_element_located(by_selector))
    el.click()
    el.send_keys(Keys.CONTROL, "a")
    el.send_keys(Keys.DELETE)
    el.clear()
    el.send_keys(value)
    return el


def pick_bank(driver, bank_name: str, timeout: int = 8):
    el = WebDriverWait(driver, timeout).until(EC.visibility_of_element_located(SELECTORS.BANK_NAME_FIELD))
    el.click()
    el.clear()
    el.send_keys(bank_name)
    time.sleep(0.8)
    try:
        option = driver.find_element(
            By.XPATH,
            f'//*[(@role="option" or self::li)][contains(., "{bank_name}")]',
        )
        option.click()
    except Exception:
        # อาจจะเป็น input ธรรมดา ไม่มี dropdown
        pass


def pick_account_type(driver, label: str, timeout: int = 8):
    el = WebDriverWait(driver, timeout).until(EC.visibility_of_element_located(SELECTORS.ACCOUNT_TYPE_FIELD))
    if el.tag_name.lower() == "select":
        Select(el).select_by_visible_text(label)
        return
    el.click()
    time.sleep(0.4)
    try:
        option = driver.find_element(
            By.XPATH,
            f'//*[(@role="option" or self::li)][contains(., "{label}")]',
        )
        option.click()
    except Exception:
        el.clear()
        el.send_keys(label)


def click_if_present(driver, by_selector, timeout: int = 3) -> bool:
    try:
        el = WebDriverWait(driver, timeout).until(EC.element_to_be_clickable(by_selector))
        el.click()
        return True
    except Exception:
        return False


# ---------- MAIN ----------
def main() -> int:
    if not EMAIL or not PASSWORD:
        print("ERROR: ตั้งค่า FLOWACCOUNT_EMAIL และ FLOWACCOUNT_PASSWORD ใน .env", file=sys.stderr)
        return 1

    print("โหลด Google Sheet...")
    rows = fetch_csv(CSV_URL)
    if len(rows) < 2:
        print("Sheet ว่างหรืออ่านไม่ได้", file=sys.stderr)
        return 1

    records = []
    for r in rows[1:]:
        # ขยาย row ให้ยาวพอ
        r = r + [""] * (20 - len(r))
        account_name = (r[0] or "").strip()       # Column A
        bank_name = (r[17] or "").strip()         # Column R
        account_number = (r[18] or "").strip()    # Column S
        if account_name and bank_name and account_number:
            records.append((account_name, bank_name, account_number))

    print(f"เจอ {len(records)} รายการที่มีข้อมูลธนาคารครบ")

    opts = Options()
    opts.add_argument("--lang=th-TH")
    opts.add_experimental_option("excludeSwitches", ["enable-automation"])
    driver = webdriver.Chrome(options=opts)
    driver.maximize_window()
    wait = WebDriverWait(driver, 30)

    try:
        # -- LOGIN --
        print("กำลัง login FlowAccount...")
        driver.get(LOGIN_URL)
        try:
            safe_fill(driver, SELECTORS.LOGIN_EMAIL, EMAIL, timeout=15)
            safe_fill(driver, SELECTORS.LOGIN_PASSWORD, PASSWORD, timeout=5)
            driver.find_element(*SELECTORS.LOGIN_SUBMIT).click()
        except Exception:
            print("Auto-login อาจไม่สำเร็จ — login เองได้ในหน้าต่างที่เปิดอยู่")

        print("รอเข้าสู่ระบบเสร็จ (รวม 2FA ถ้ามี)... สูงสุด 3 นาที")
        WebDriverWait(driver, 180).until(lambda d: "flowaccount.com/N" in d.current_url)

        ok, failed = [], []
        for account_name, bank_name, account_number in records:
            print(f"\n→ {account_name} | {bank_name} | {account_number}")
            try:
                driver.get(CONTACTS_URL)
                search = wait.until(EC.visibility_of_element_located(SELECTORS.CONTACT_SEARCH))
                search.click()
                search.clear()
                search.send_keys(account_name)
                time.sleep(1.5)

                # คลิก row แรกที่ตรงชื่อ
                target = wait.until(EC.element_to_be_clickable((
                    By.XPATH,
                    f'(//tr[contains(., "{account_name}")] | //*[@role="row"][contains(., "{account_name}")] | //a[contains(., "{account_name}")])[1]',
                )))
                target.click()
                time.sleep(1.0)

                click_if_present(driver, SELECTORS.EDIT_BUTTON, timeout=4)
                click_if_present(driver, SELECTORS.BANK_TAB, timeout=4)
                click_if_present(driver, SELECTORS.ADD_BANK_BUTTON, timeout=2)

                pick_bank(driver, bank_name)
                safe_fill(driver, SELECTORS.ACCOUNT_NAME_FIELD, account_name)
                safe_fill(driver, SELECTORS.ACCOUNT_NUMBER_FIELD, account_number)
                safe_fill(driver, SELECTORS.BRANCH_CODE_FIELD, BRANCH_CODE)
                safe_fill(driver, SELECTORS.BRANCH_NAME_FIELD, BRANCH_NAME)
                pick_account_type(driver, ACCOUNT_TYPE)

                driver.find_element(*SELECTORS.SAVE_AND_CLOSE).click()
                time.sleep(1.5)
                print("  ✓ บันทึกแล้ว")
                ok.append(account_name)
            except Exception as exc:
                print(f"  ✗ ล้มเหลว: {exc}")
                failed.append((account_name, str(exc)))
                try:
                    driver.save_screenshot(f"error_{int(time.time())}.png")
                except Exception:
                    pass

        print("\n========== สรุป ==========")
        print(f"สำเร็จ: {len(ok)}")
        print(f"ล้มเหลว: {len(failed)}")
        for name, err in failed:
            print(f"  - {name}: {err}")
    finally:
        driver.quit()

    return 0


if __name__ == "__main__":
    sys.exit(main())
